ksql-engine/src/main/java/io/confluent/ksql/util/QueryMetadata.java

```diff
@@ -68,8 +79,24 @@ public String getQueryApplicationId() {
   }
 
   public void close() {
-    kafkaStreams.close(100L, TimeUnit.MILLISECONDS);
-    kafkaStreams.cleanUp();
+    kafkaStreams.close();
+    long pollStart = System.currentTimeMillis();
+    long pollEnd = pollStart + (long) ksqlConfig.get(KsqlConfig.KSQL_QUERY_CLOSE_WAIT_TIME_CONFIG);
+    while (System.currentTimeMillis() < pollEnd
+           && kafkaStreams.state() != KafkaStreams.State.NOT_RUNNING) {
+      try {
+        Thread.sleep(1000);
```

Comment:
What is the context where this code will run? I am worried about the 1 second sleeps, blocking other operations.

Reply:
This will run when we terminate a query. It won't affect execution of other queries.
We can change it to lower time if necessary.

Reply:
if you do `kafkaStreams.close()` without a timeout it will block until it has closed. So the loop would not be needed.